### ISSUE TITLE
Not able to 'get' ButteRFly interface

### ISSUE THREAD

**Issue by AOShei** [RCayre/injectable-firmware#1]
I've loaded up the firmware for the nordic nRF but none of the examples work I keep getting this error:
```
[INFO] Module ble_sniff loaded !
Traceback (most recent call last):
  File "./mirage_launcher", line 4, in <module>
    main()
  File "/home/andrew/mirage/mirage/mirage.py", line 11, in main
    parser.run()
  File "/home/andrew/mirage/mirage/core/argParser.py", line 141, in run
    self.launcher()
  File "/home/andrew/mirage/mirage/core/argParser.py", line 120, in launcher
    self.appInstance.run()
  File "/home/andrew/mirage/mirage/core/app.py", line 327, in run
    output = module["module"].execute()
  File "/home/andrew/mirage/mirage/core/module.py", line 109, in execute
    output = self.run()
  File "/home/andrew/mirage/mirage/modules/ble_sniff.py", line 300, in run
    self.initEmittersAndReceivers()
  File "/home/andrew/mirage/mirage/modules/ble_sniff.py", line 183, in initEmittersAndReceivers
    self.emitters.append(self.getEmitter(interface=interface))
  File "/home/andrew/mirage/mirage/core/module.py", line 244, in getEmitter
    self.__class__.Emitters[interface] = self.__class__.EmittersClass[self.technology](interface=interface)
  File "/home/andrew/mirage/mirage/libs/ble.py", line 494, in __init__
    super().__init__(interface=interface, packetType=BLEPacket, deviceType=deviceClass)
  File "/home/andrew/mirage/mirage/libs/wireless.py", line 31, in __init__
    self.device = self.deviceType.get(self.interface)
AttributeError: 'NoneType' object has no attribute 'get'

```

Additional notes: When I run lsusb the device appears with no name
```
Bus 001 Device 025: ID 5a17:0000 
```

I've re-flashed the firmware three times now with no noticeable incident but same result.

OS: Ubuntu 20.04

**Comment by RCayre** [comment]
Hi,
Weird issue. Do you have another nrf52840 dongle available for testing? Can you also provide me the output of dmesg after the dongle was plugged in?

**Comment by AOShei** [comment]
Hello,

This is what I'm logging when plugging in the device, note the USB hub is necessary as my computer only has type-c ports. I'm going to see if I can test it with another one.
```
[10448.057003] usb 1-9: new high-speed USB device number 11 using xhci_hcd
[10448.212328] usb 1-9: New USB device found, idVendor=05e3, idProduct=0610, bcdDevice= 6.54
[10448.212331] usb 1-9: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[10448.212332] usb 1-9: Product: USB2.1 Hub
[10448.212334] usb 1-9: Manufacturer: GenesysLogic
[10448.213119] hub 1-9:1.0: USB hub found
[10448.213354] hub 1-9:1.0: 4 ports detected
[10448.525262] usb 1-9.3: new full-speed USB device number 12 using xhci_hcd
[10448.652888] usb 1-9.3: New USB device found, idVendor=5a17, idProduct=0000, bcdDevice= 0.01
[10448.652894] usb 1-9.3: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[10448.652898] usb 1-9.3: Product: ButteRFly
[10448.652901] usb 1-9.3: Manufacturer: Mirage Toolkit
[10450.649434] usb 4-1: new SuperSpeed Gen 1 USB device number 3 using xhci_hcd
[10450.675763] usb 4-1: New USB device found, idVendor=05e3, idProduct=0626, bcdDevice= 6.54
[10450.675769] usb 4-1: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[10450.675773] usb 4-1: Product: USB3.1 Hub
[10450.675776] usb 4-1: Manufacturer: GenesysLogic
[10450.678409] hub 4-1:1.0: USB hub found
[10450.678829] hub 4-1:1.0: 4 ports detected
```
It appears to register correctly at least at first.

One thing to note, when I compiled the firmware I used the most recent version of the GNU Arm Tools. Though, I noticed in the nord nRF5 SDK it calls for version 7-2018-q2-update. I modified the nord SDK to look for the newer version as the older version does not appear to be available anymore... But like I said it compiles without error and has no problem flashing the dongle.

Maybe its a udev or permissions problem? Idk just a shot in the dark. I'm not an expert on how linux kernel creates new device interfaces. lsusb still returns like before and I've tried searching for a dev address so I can enter this manually into mirage but I can't find it, maybe I'm looking in the wrong place though.

Unfortunately I do not have another dongle, I just ordered one but it'll be a week or so until I get it.

Hopefully you have a better idea of where to look than I do, but based on the dmesg log I think the issue might not be the firmware. Any help you can offer is greatly appreciated though as I was hoping to use this as a test case for a university research project I'm doing.

**Comment by AOShei** [comment]
Hi,

I have a quick update. I just tried connecting with another hub as I noticed the other one was usb 2.1 and maybe was creating a conflict. Now I get this:

```
dmesg output:
[12098.810646] usb 1-9: USB disconnect, device number 14
[12104.335192] usb 1-9: new full-speed USB device number 15 using xhci_hcd
[12104.484332] usb 1-9: New USB device found, idVendor=5a17, idProduct=0000, bcdDevice= 0.01
[12104.484334] usb 1-9: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[12104.484335] usb 1-9: Product: ButteRFly
[12104.484336] usb 1-9: Manufacturer: Mirage Toolkit
```
```
lsusb output:
Bus 001 Device 015: ID 5a17:0000 Mirage Toolkit ButteRFly
```

However, when running the examples I still have: 
```
AttributeError: 'NoneType' object has no attribute 'get'
```

**Comment by RCayre** [comment]
Thanks for the additional details. I guess the problem is linked to a Mirage bug or a library issue.
Can you give me the output of the following python script:

```
import usb

# ButterRFly USB device Identifiers
BUTTERFLY_ID_VENDOR = 0x5A17
BUTTERFLY_ID_PRODUCT = 0x0000

devicesList = list(usb.core.find(idVendor=BUTTERFLY_ID_VENDOR, idProduct=BUTTERFLY_ID_PRODUCT,find_all=True))
print(devicesList)
if len(devicesList) > 0:
	dongle = devicesList[0]
	dongle.set_configuration()
	dongle.write(0x01, bytes.fromhex("5a17000000"))
	print(dongle.read(0x81,64))
```

**Comment by AOShei** [comment]
Yup, I get
```
[<DEVICE ID 5a17:0000 on Bus 001 Address 007>]
array('B', [90, 23, 1, 0, 0, 1, 0])
```

**Comment by RCayre** [comment]
It's really weird. You seems to be able to communicate with the dongle and got the expected response.
Did you run `sudo python3 setup.py install` ? can you try without ./mirage_launcher, by directly launching the mirage command ? 
For example `sudo mirage ble_info INTERFACE=butterfly0` ?

**Comment by AOShei** [comment]
Unfortunately I get the same result.

**Comment by AOShei** [comment]
Question, is the ButteRFly meant to appear as a serial interface?

**Comment by AOShei** [comment]
Hello,
So I'm going to close this issue turns out the problem was Mirage, well sort of.
When I originally downloaded Mirage I was following the instructions at laas.fr [here](https://homepages.laas.fr/rcayre/mirage-documentation/install.html) which provides the address for the git repo at laas.fr. Turns out that repo is outdated.
I replaced it with the Github Mirage-Master repo and I'm up and running.
So, to avoid anyone else running into the same problem, I would suggest changing the help doc so the install instructions point to the Github repo instead, or merging the changes with the repo at laas.fr.
Thank you for taking the time to respond!
Have a good evening.
